### ISSUE TITLE
TFLint Plugins

### ISSUE THREAD

**Issue by starlightromero** [renovatebot/renovate#14906]
### What would you like Renovate to be able to do?

I would like Renovate to be able to update TFLint plugin versions. TFLint plugins, by convention, are located in a `.tflint.hcl` file. TFLint plugins include, but are not limited to:

- [AWS](https://github.com/terraform-linters/tflint-ruleset-aws)
```hcl
plugin "aws" {
    enabled = true
    version = "0.13.2"
    source  = "github.com/terraform-linters/tflint-ruleset-aws"
}
```

- [Google](https://github.com/terraform-linters/tflint-ruleset-google)
```hcl
plugin "google" {
    enabled = true
    version = "0.16.1"
    source  = "github.com/terraform-linters/tflint-ruleset-google"
}
```

- [Azure](https://github.com/terraform-linters/tflint-ruleset-azurerm)
```hcl
plugin "azurerm" {
    enabled = true
    version = "0.15.0"
    source  = "github.com/terraform-linters/tflint-ruleset-azurerm"
}
```

### If you have any ideas on how this should be implemented, please tell us here.

No ideas currently. I'm not the strongest in TypeScript but I'd be willing to help implement

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by starlightromero** [comment]
# New package manager questionnaire

**Did you read our documentation on adding a package manager?**

- [x] I've read the [adding a package manager](adding-a-package-manager.md) documentation.

## Basics

### Name of package manager

There is no official package manager for TFLint.

### What language does this support?

TFLint configuration is written in HCL.

### How popular is this package manager?

There is no official package manager for TFLint. However TFLint is fairly popular - 3k stars on GitHub.

### Does this language have other (competing?) package managers?

- [ ] Yes (give names)
- [x] No

---

## Package File Detection

### What type of package files and names does it use?

`.tflint.hcl`

### What [fileMatch](https://renovatebot.com/docs/configuration-options/#filematch) pattern(s) should be used?

`.tflint.hcl`

### Is it likely that many users would need to extend this pattern for custom file names?

- [ ] Yes
- [x] No

### Is the fileMatch pattern likely to get many "false hits" for files that have nothing to do with package management?

No, it is an exact match.

---

## Parsing and Extraction

### Can package files have "local" links to each other that need to be resolved?

No.

### Is there a reason why package files need to be parsed together (in serial) instead of independently?

No.

### What format/syntax is the package file in?

- [ ] JSON
- [ ] TOML
- [ ] YAML
- [x] Custom (explain below)

HCL

### How do you suggest parsing the file?

- [ ] Off the shelf parser
- [x] Using regex
- [ ] Custom-parsed line by line
- [ ] Other

### Does the package file structure distinguish between different "types" of dependencies? e.g. production dependencies, dev dependencies, etc?

- [ ] Yes, production and development dependencies
- [x] No, all dependencies are treated the same

### List all the sources/syntaxes of dependencies that can be extracted

[AWS](https://github.com/terraform-linters/tflint-ruleset-aws)
plugin "aws" {
    enabled = true
    version = "0.13.2"
    source  = "github.com/terraform-linters/tflint-ruleset-aws"
}

[Google](https://github.com/terraform-linters/tflint-ruleset-google)
plugin "google" {
    enabled = true
    version = "0.16.1"
    source  = "github.com/terraform-linters/tflint-ruleset-google"
}

[Azure](https://github.com/terraform-linters/tflint-ruleset-azurerm)
plugin "azurerm" {
    enabled = true
    version = "0.15.0"
    source  = "github.com/terraform-linters/tflint-ruleset-azurerm"
}

### Describe which types of dependencies above are supported and which will be implemented in future

The ones listed above are supported. There may be other custom plugins. But the ones listed above are the most used.

---

## Versioning

### What versioning scheme does the package file(s) use?

SemVer

### Does this versioning scheme support range constraints, e.g. `^1.0.0` or `1.x`?

- [ ] Supports range constraints (e.g `^1.0.0` or `1.x`)
- [x] No

### Is this package manager used for applications, libraries, or both? If both, is there a way to tell which is which?

- [ ] Applications
- [x] Libraries
- [ ] Both (explain how to tell which is which)

### If ranges are supported, are there any cases when Renovate should pin ranges to exact versions if rangeStrategy=auto?

N/A. Ranges are not supported.

---

## Lookup

### Is a new datasource required? Provide details

- [x] Yes, provide details.
- [ ] No.

The datasource will need to be the GitHub releases of the TFLint plugins [AWS](https://github.com/terraform-linters/tflint-ruleset-aws/releases), [Google](https://github.com/terraform-linters/tflint-ruleset-google/releases), [Azure](https://github.com/terraform-linters/tflint-ruleset-azurerm/releases).

### Will users need the capability to specify a custom host/registry to look up? Can it be found within the package files, or within other files inside the repository, or would it require Renovate configuration?

Users will not need to specify a custom host/registry.

### Do the package files have any "constraints" on the parent language (e.g. supports only v3.x of Python) or platform (Linux, Windows, etc) that should be used in the lookup procedure?

No.

### Will users need the ability to configure language or other constraints using Renovate config?

No.

---

## Artifacts

### Are lock files or checksum files used? Are they mandatory?

These are not used.

### If so, what tool and exact commands should be used if updating one or more package versions in a dependency file?

N/A

### If applicable, describe how the tool maintains a cache and if it can be controlled via CLI or env? Do you recommend the cache be kept or disabled/ignored?

The tools maintains a cache via the specified [plugins directory](https://github.com/terraform-linters/tflint/blob/master/docs/user-guide/plugins.md#plugin-directory.)

### If applicable, what command should be used to generate a lock file from scratch if you already have a package file? This will be used for "lock file maintenance"

N/A

## Other

### Is there anything else to know about this package manager?

Nothing else comes to mind.

**Comment by rarkins** [comment]
If it's using github releases, I think we can just reuse our existing github-releases datasource? BTW are you sure it's only releases and not tags or git refs?

**Comment by viceice** [comment]
this looks like it can be easily handled by the terraform manager? as those plugins are for terraform lint. 🤔

**Comment by rarkins** [comment]
Seems to be its own tool/package manager though: https://github.com/terraform-linters/tflint

**Comment by viceice** [comment]
ok

**Comment by starlightromero** [comment]
> If it's using github releases, I think we can just reuse our existing github-releases datasource? BTW are you sure it's only releases and not tags or git refs?

It uses [tags](https://github.com/terraform-linters/tflint-ruleset-aws/tags) which are inline with the [releases](https://github.com/terraform-linters/tflint-ruleset-aws/releases). They both point to the same git ref. For example, AWS plugin release v0.13.2 and tag v0.13.2 both point to ref `87f78762c3273551274d7c56e11353bc08d137ac`.

**Comment by rarkins** [comment]
They are not always the same. You can have a tag without having a release. Sounds like we should use GitHub-tags datasource, not releases.

**Comment by rarkins** [comment]
It would be releases if tflint depends on being able to download release assets instead of git source code.

**Comment by starlightromero** [comment]
> They are not always the same. You can have a tag without having a release.
Correct. I was more-so saying for these plugins specified, they have tags and releases with the same git ref. 

> It would be releases if tflint depends on being able to download release assets instead of git source code.
I'm not sure if tflint depends on those release assets or not. Either way I think the GitHub-tag datasource is a good fit like you say.

**Comment by multani** [comment]
@starlightromero I think this is now implemented by #17954.
